File: file-system/src/main/java/com/storyteller_f/file_system/FileInstanceFactory.java
package com.storyteller_f.file_system;

import android.content.Context;
import android.os.Build;
import android.os.StatFs;
import android.util.Log;


import com.storyteller_f.file_system.instance.local.EmulatedLocalFileInstance;
import com.storyteller_f.file_system.instance.local.ExternalLocalFileInstance;
import com.storyteller_f.file_system.instance.local.LocalFileInstance;
import com.storyteller_f.file_system.instance.local.document.ExternalDocumentLocalFileInstance;
import com.storyteller_f.file_system.instance.local.document.MountedLocalFileInstance;
import com.storyteller_f.file_system.model.DirectoryItemModel;
import com.storyteller_f.file_system.model.FileItemModel;

import java.util.ArrayList;
import java.util.List;

public class FileInstanceFactory {
    private static final String TAG = "FileInstanceFactory";
    static final String rootUserEmulatedPath = "/storage/emulated/0/";
    static final String emulatedRootPath = "/storage/emulated/";
    static final String storagePath = "/storage/";
    private static final Filter filter = new Filter() {
        @Override
        public boolean onPath(String parent, String absolutePath, boolean isFile) {
            return true;
        }

        @Override
        public List<FileItemModel> onFile(String parent) {
            return new ArrayList<>();
        }

        @Override
        public List<DirectoryItemModel> onDirectory(String parent) {
            return new ArrayList<>();
        }
    };

    public static LocalFileInstance getFileInstance(String path, Context context) {
        return getFileInstance(filter, path, context);
    }

    public static LocalFileInstance getFileInstance(Filter filter, String path, Context context) {
//        if (path.endsWith("//")) {
//            Log.e(TAG, "getFileInstance: path末尾是//:"+path);
//        }
        if (path.startsWith(rootUserEmulatedPath)) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R && Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
                return new ExternalDocumentLocalFileInstance(filter, context, path);
            } else {
                return new ExternalLocalFileInstance(context, filter, path);
            }
        } else if (path.startsWith(emulatedRootPath)) {
            return new EmulatedLocalFileInstance(context, filter);
        } else if (path.startsWith("/storage/self/")) {
            return new ExternalLocalFileInstance(context, filter, path);
        } else if (path.equals(storagePath)) {
            return new ExternalLocalFileInstance(context, filter, path);
        } else if (path.startsWith(storagePath)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                return new ExternalLocalFileInstance(context, filter, path);
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                return new MountedLocalFileInstance(filter, context, path);
            } else if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN_MR1) {
                Log.e(TAG, "getFileInstance: 当前版本不支持formTreeUri，测试性的返回");
                return new ExternalLocalFileInstance(context, filter, path);
            } else {
                return new ExternalLocalFileInstance(context, filter, path);
            }

        } else {
            return new ExternalLocalFileInstance(context, filter, path);
        }
    }

    public static String getPrefix(String path) {
        if (path.startsWith(rootUserEmulatedPath)) {
            return rootUserEmulatedPath;
        } else if (path.startsWith(emulatedRootPath)) {
            return emulatedRootPath;
        } else if (path.equals(storagePath)) {
            return "/";
        } else if (path.startsWith(storagePath)) {
            int endIndex = path.indexOf("/", storagePath.length());
            if (endIndex == -1) endIndex = path.length();
            return path.substring(0, endIndex + 1);
        } else {
            return "/";
        }
    }

    public static long getSpace(String prefix) {
        StatFs stat = new StatFs(prefix);
        long blockSize;
        long availableBlocks;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            blockSize = stat.getBlockSizeLong();
            availableBlocks = stat.getAvailableBlocksLong();
        } else {
            blockSize = stat.getBlockSize();
            availableBlocks = stat.getAvailableBlocks();
        }
        return blockSize * availableBlocks;
    }
}

File: file-system/src/main/java/com/storyteller_f/file_system/FileInstanceFactory.kt
package com.storyteller_f.file_system

import android.content.Context
import android.os.Build
import android.os.StatFs
import android.util.Log
import com.storyteller_f.file_system.model.FileItemModel
import com.storyteller_f.file_system.model.DirectoryItemModel
import com.storyteller_f.file_system.instance.FileInstance
import com.storyteller_f.file_system.instance.local.document.ExternalDocumentLocalFileInstance
import com.storyteller_f.file_system.instance.local.RegularLocalFileInstance
import com.storyteller_f.file_system.instance.local.fake.EmulatedLocalFileInstance
import com.storyteller_f.file_system.instance.local.fake.StorageLocalFileInstance
import com.storyteller_f.file_system.instance.local.document.MountedLocalFileInstance
import com.storyteller_f.file_system.instance.local.fake.FakeDirectoryLocalFileInstance
import java.lang.Exception
import java.util.ArrayList

object FileInstanceFactory {
    private const val TAG = "FileInstanceFactory"
    const val rootUserEmulatedPath = "/storage/emulated/0"
    const val emulatedRootPath = "/storage/emulated"
    const val currentEmulatedPath = "/storage/self"
    const val storagePath = "/storage"
    const val sdCardPath = "/sdcard"
    private val filter: Filter = object : Filter {
        override fun onPath(parent: String, absolutePath: String, isFile: Boolean): Boolean {
            return true
        }

        override fun onFile(parent: String): List<FileItemModel> {
            return ArrayList()
        }

        override fun onDirectory(parent: String): List<DirectoryItemModel> {
            return ArrayList()
        }
    }

    @JvmStatic
    fun getFileInstance(path: String, context: Context): FileInstance {
        return getFileInstance(filter, path, context)
    }

    private fun getFileInstance(filter: Filter?, path: String, context: Context): FileInstance {
        assert(!path.endsWith("/"))
        return when {
            path.startsWith(rootUserEmulatedPath) -> if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R && Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
                ExternalDocumentLocalFileInstance(filter, context, path)
            } else {
                RegularLocalFileInstance(context, filter, path)
            }
            path.startsWith(emulatedRootPath) -> EmulatedLocalFileInstance(context, filter)
            path.startsWith(currentEmulatedPath) -> RegularLocalFileInstance(context, filter, path)
            path == storagePath -> StorageLocalFileInstance(context)
            path.startsWith(storagePath) -> when {
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.R ->
                    RegularLocalFileInstance(context, filter, path)
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP ->
                    MountedLocalFileInstance(filter, context, path)
                Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN_MR1 -> {
                    Log.e(
                        TAG,
                        "getFileInstance: 当前版本不支持formTreeUri，测试性的返回"
                    )
                    RegularLocalFileInstance(context, filter, path)
                }
                else -> RegularLocalFileInstance(context, filter, path)
            }
            path.startsWith(sdCardPath) -> RegularLocalFileInstance(context, filter, path)
            path.startsWith("/mnt") -> RegularLocalFileInstance(context, filter, path)
            path.startsWith("/data/data/${context.packageName}") -> RegularLocalFileInstance(context, filter, path)
            else -> FakeDirectoryLocalFileInstance(path, context)
        }
    }

    @JvmStatic
    fun getPrefix(path: String, context: Context): String {
        assert(!path.endsWith("/")) {
            path
        }
        return when {
            path.startsWith(rootUserEmulatedPath) -> rootUserEmulatedPath
            path.startsWith(emulatedRootPath) -> emulatedRootPath
            path == storagePath -> "/storage"
            path.startsWith(storagePath) -> {
                var endIndex = path.indexOf("/", storagePath.length)
                if (endIndex == -1) endIndex = path.length
                path.substring(0, endIndex + 1)
            }
            path.startsWith(sdCardPath) -> sdCardPath
            path.startsWith("/mnt") -> "/mnt"
            path.startsWith("/data/data/${context.packageName}") -> "/data/data/${context.packageName}"
            else -> ""
        }
    }

    fun getSpace(prefix: String?): Long {
        val stat = StatFs(prefix)
        val blockSize: Long
        val availableBlocks: Long
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            blockSize = stat.blockSizeLong
            availableBlocks = stat.availableBlocksLong
        } else {
            blockSize = stat.blockSize.toLong()
            availableBlocks = stat.availableBlocks.toLong()
        }
        return blockSize * availableBlocks
    }

    @Throws(Exception::class)
    fun toChild(
        fileInstance: FileInstance,
        name: String,
        isFile: Boolean,
        context: Context
    ): FileInstance {
        val parentPrefix = getPrefix(fileInstance.path, context)
        val path = fileInstance.path + "/" + name
        val childPrefix = getPrefix(path, context)
        return if (parentPrefix == childPrefix) {
            fileInstance.toChild(name, isFile, false)
        } else {
            getFileInstance(path, context)
        }
    }

    @Throws(Exception::class)
    fun toParent(fileInstance: FileInstance, context: Context): FileInstance {
        val parentPrefix = getPrefix(fileInstance.parent, context)
        val childPrefix = getPrefix(fileInstance.path, context)
        return if (parentPrefix == childPrefix) {
            fileInstance.toParent()
        } else {
            getFileInstance(fileInstance.parent, context)
        }
    }
}